File: ios/Godstone/Sources/GodstoneMesh/TrustedPeerChannel.swift
import Foundation

// ---------------------------------------------------------------------------
// T24 - the peer-event contract (the iOS half, mirror of TrustedPeerChannel.kt).
//
// A peer is publish'd to the application onely as a relation-bound, authenticated
// view. Its identity is carri'd by the node id and the identity public key and
// the trust version - never by a transport address, MAC, CBPeripheral UUID, node
// hint or RSSI, which may not substitute for it (section 5, the frozen identity
// law: node_id = BLAKE2s-128(identityPub), 16 octets; identityPub, Ed25519, 32
// octets). The events travel a BOUNDED RELIABLE channel, never a silent-drop
// queue: an offer is accepted, meeteth bounded backpressure (the caller keepeth
// the duty to retry), or the channel is in terminal failure. Delivery followeth
// the section 13 event algorithm - capture the owner token, validate the token
// and the input under the named owner, make one transition, then revalidate the
// token upon completion.
// ---------------------------------------------------------------------------

/// The frozen widths of an authenticated peer identity.
enum TrustedPeerSpec {
    static let nodeIdBytes: Int = 16
    static let identityPubBytes: Int = 32
}

/// An immutable, authenticated view of one peer bound to exactly one relation.
/// Two peers are the selfsame peer when their relation, node id, identity public
/// key and trust version all agree by CONTENT - the transport handle is no part
/// of identity. As a value type it is copied by value, so no caller may mutate
/// a peer it hath been given.
struct TrustedPeer: Equatable, Hashable, Sendable {
    let relation: RelationKey
    let nodeId16: Data
    let identityPub32: Data
    let trustVersion: Int

    /// A failable init: a peer whose node id is not sixteen octets, whose
    /// identity public key is not thirty-two octets, or whose trust version is
    /// negative is REFUS'd (nil), never constructured.
    init?(relation: RelationKey, nodeId16: Data, identityPub32: Data, trustVersion: Int) {
        self.relation = relation
        self.nodeId16 = nodeId16
        self.identityPub32 = identityPub32
        self.trustVersion = trustVersion
        guard nodeId16.count == TrustedPeerSpec.nodeIdBytes else { return nil }
        guard identityPub32.count == TrustedPeerSpec.identityPubBytes else { return nil }
        guard trustVersion >= 0 else { return nil }
    }

    // Content equality & hash are synthesiz'd from the four stored properties:
    // Data compareth bytewise, RelationKey is Hashable, Int compareth by value -
    // therefore two independently-built peers with the selfsame bytes are equal,
    // while a differing byte, version or relation maketh a different peer.
}

/// The three authoritative peer events, each carrying its immutable trusted peer.
enum LinkEvent: Sendable {
    /// The relation became link-ready: trusted cryptographic ready and key-confirmed.
    case linkReady(TrustedPeer)
    /// The relation was lost or torn down.
    case linkLost(TrustedPeer)
    /// The peer's authenticated identity was assert'd at its trust version.
    case auth(TrustedPeer)

    var peer: TrustedPeer {
        switch self {
        case let .linkReady(peer), let .linkLost(peer), let .auth(peer):
            return peer
        }
    }
}

/// The verdict a caller readeth back from offering upon the reliable channel.
enum OfferVerdict: Equatable {
    /// The event was carri'd by the channel.
    case accepted
    /// Bounded: the queue is full. Nothing is silently dropped; the caller must retry.
    case backpressure
    /// The channel, or the owning relation, hath fail'd terminally; nothing more is carri'd.
    case terminalFailure
}

/// A bounded RELIABLE peer-event channel. It is never lossy by silent drop: an
/// offer is accepted, meeteth bounded backpressure (the caller keepeth the duty
/// to retry), or the channel is in terminal failure. The owner token (the
/// relation generation) is validated at entry and revalidated upon completion,
/// and exactly one transition is made per accepted offer.
final class ReliablePeerEventChannel: @unchecked Sendable {
    let capacity: Int
    private let lock = NSLock()
    private var queue: [LinkEvent] = []
    private var terminal: Bool = false

    init(capacity: Int) {
        precondition(capacity > 0, "the bound of a reliable channel must be positive")
        self.capacity = capacity
    }

    /// How many events await delivery.
    var size: Int {
        lock.lock(); defer { lock.unlock() }
        return queue.count
    }

    /// Whether the channel hath fail'd terminally.
    var isTerminal: Bool {
        lock.lock(); defer { lock.unlock() }
        return terminal
    }

    /// Offer [event] under the named [ownerGeneration]. A terminal channel, or a
    /// token that doth not match the event's own relation generation, is refus'd
    /// as a terminal failure; a full queue answereth backpressure without
    /// dropping a whit; otherwise the single event is accepted.
    func offer(_ event: LinkEvent, ownerGeneration: UInt64) -> OfferVerdict {
        lock.lock(); defer { lock.unlock() }
        if terminal { return .terminalFailure }
        if event.peer.relation.generation != ownerGeneration { return .terminalFailure }
        if queue.count >= capacity { return .backpressure }
        queue.append(event)
        // revalidate the owner token upon completion: a mid-offer terminal turn undoeth the slip
        if terminal {
            if let last = queue.last, last.peer == event.peer { queue.removeLast() }
            return .terminalFailure
        }
        return .accepted
    }

    /// Take the head event for delivery, or nil when empty or terminal.
    func poll() -> LinkEvent? {
        lock.lock(); defer { lock.unlock() }
        if terminal { return nil }
        return queue.isEmpty ? nil : queue.removeFirst()
    }

    /// Fail the channel terminally: nothing more is carri'd, and the queue is voided.
    func failTerminally() {
        lock.lock(); defer { lock.unlock() }
        terminal = true
        queue.removeAll()
    }
}

File: android/mesh/src/main/java/io/godstone/mesh/transport/TrustedPeerChannel.kt
package io.godstone.mesh.transport

// ---------------------------------------------------------------------------
// T24 - the peer-event contract.
//
// A peer is published to the application ONLY as a relation-bound,
// authenticated view. Its identity is carried by the node id and the identity
// public key and the trust version - never by a transport address, MAC,
// CBPeripheral UUID, node hint or RSSI, which may not substitute for it
// (section 5, the frozen identity law: node_id = BLAKE2s-128(identityPub),
// 16 bytes; identityPub, Ed25519, 32 bytes). The events travel a BOUNDED
// RELIABLE channel, never a silent-drop queue: an offer is Accepted, meets
// bounded Backpressure (the caller keeps the duty to retry), or the channel is
// in TerminalFailure. Delivery follows the section 13 event algorithm -
// capture the owner token, validate the token and the input under the named
// owner, make one transition, then revalidate the token upon completion.
// ---------------------------------------------------------------------------

/** The frozen widths of an authenticated peer identity. */
internal object TrustedPeerSpec {
    const val NODE_ID_BYTES: Int = 16
    const val IDENTITY_PUB_BYTES: Int = 32
}

/**
 * An immutable, authenticated view of one peer bound to exactly one relation.
 * Two peers are the same peer when their relation, node id, identity public
 * key and trust version all agree by content - the transport handle is not
 * part of identity.
 */
internal class TrustedPeer(
    val relation: RelationKey,
    suppliedNodeId: ByteArray,
    suppliedIdentityPub: ByteArray,
    val trustVersion: Long,
) {
    // The captured identity is IMMUTABLE: the value owneth private copies of the
    // authenticated bytes, so no later mutation of a caller's buffer may corrupt a
    // peer that hath already been trusted (the section-5 immutability law).
    val nodeId16: ByteArray = suppliedNodeId.copyOf()
    val identityPub32: ByteArray = suppliedIdentityPub.copyOf()

    init {
        require(nodeId16.size == TrustedPeerSpec.NODE_ID_BYTES) {
            "a trusted peer carrieth a ${TrustedPeerSpec.NODE_ID_BYTES}-byte node id"
        }
        require(identityPub32.size == TrustedPeerSpec.IDENTITY_PUB_BYTES) {
            "a trusted peer carrieth a ${TrustedPeerSpec.IDENTITY_PUB_BYTES}-byte identity public key"
        }
        require(trustVersion >= 0L) { "a trust version may not be negative" }
    }

    companion object {
        /**
         * Capture the authenticated relation as an immutable identity AT THE
         * AUTHENTICATED MOMENT, forcing the frozen identity law upon the capture:
         * the node id is DERIVED as `BLAKE2s-128(identityPub)` by the very routine
         * the identity layer employs (section 5), never trusted from a caller's
         * bytes. The raw constructor accepteth any node id the caller supplyeth;
         * this factory leaveth no such freedom, so a captured peer can never
         * disagree with its own identity public key.
         *
         * @param authenticatedIdentityPub32 the remote Ed25519 identity public key
         *        (32 bytes) as the transport learned it when the relation was sealed.
         * @param trustVersion a NON-NEGATIVE, monotonic trust epoch of that identity.
         */
        fun capture(
            relation: RelationKey,
            authenticatedIdentityPub32: ByteArray,
            trustVersion: Long,
        ): TrustedPeer {
            require(authenticatedIdentityPub32.size == TrustedPeerSpec.IDENTITY_PUB_BYTES) {
                "onely a ${TrustedPeerSpec.IDENTITY_PUB_BYTES}-byte authenticated identity public key may be captured"
            }
            require(trustVersion >= 0L) { "a trust version may not be negative" }
            // Defer wholly to the canonical derivation; duplicate no hash here.
            val nodeId16 = io.godstone.mesh.identity.Identity.nodeIdOf(authenticatedIdentityPub32)
            return TrustedPeer(relation, nodeId16, authenticatedIdentityPub32, trustVersion)
        }
    }

    /** A defensive read of the node id, that no caller may mutate our copy. */
    fun copyNodeId(): ByteArray = nodeId16.copyOf()

    /** A defensive read of the identity public key. */
    fun copyIdentityPub(): ByteArray = identityPub32.copyOf()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is TrustedPeer) return false
        return other.relation == relation &&
            other.trustVersion == trustVersion &&
            other.nodeId16.contentEquals(nodeId16) &&
            other.identityPub32.contentEquals(identityPub32)
    }

    override fun hashCode(): Int {
        var h = relation.hashCode()
        h = 31 * h + nodeId16.contentHashCode()
        h = 31 * h + identityPub32.contentHashCode()
        h = 31 * h + trustVersion.hashCode()
        return h
    }
}

/** The three authoritative peer events, each carrying its immutable trusted peer. */
internal sealed class LinkEvent {
    abstract val peer: TrustedPeer

    /** The relation became link-ready: trusted cryptographic ready and key-confirmed. */
    internal class LinkReady(override val peer: TrustedPeer) : LinkEvent()

    /** The relation was lost or torn down. */
    internal class LinkLost(override val peer: TrustedPeer) : LinkEvent()

    /** The peer's authenticated identity was asserted at its trust version. */
    internal class Auth(override val peer: TrustedPeer) : LinkEvent()
}

/** The verdict a caller readeth back from offering upon the reliable channel. */
internal sealed class OfferVerdict {
    /** The event was carried by the channel. */
    internal object Accepted : OfferVerdict()

    /** Bounded: the queue is full. Nothing is silently dropped; the caller must retry. */
    internal object Backpressure : OfferVerdict()

    /** The channel, or the owning relation, hath fail'd terminally; nothing more is carried. */
    internal object TerminalFailure : OfferVerdict()
}

/**
 * A bounded RELIABLE peer-event channel. It is never lossy by silent drop: an
 * offer is Accepted, meeteth bounded Backpressure (the caller keepeth the duty
 * to retry), or the channel is in TerminalFailure. The owner token (the
 * relation generation) is validated at entry and revalidated upon completion,
 * and exactly one transition is made per accepted offer.
 */
internal class ReliablePeerEventChannel(val capacity: Int) {
    init {
        require(capacity > 0) { "the bound of a reliable channel must be positive" }
    }

    private val queue = java.util.ArrayDeque<LinkEvent>()
    private var terminal = false

    /** How many events await delivery. */
    val size: Int get() = queue.size

    /** Whether the channel hath fail'd terminally. */
    val isTerminal: Boolean get() = terminal

    /**
     * Offer [event] under the named [ownerGeneration]. A terminal channel, or a
     * token that doth not match the event's own relation generation, is refused
     * as a [OfferVerdict.TerminalFailure]; a full queue answereth
     * [OfferVerdict.Backpressure] without dropping a whit; otherwise the single
     * event is [OfferVerdict.Accepted].
     */
    fun offer(event: LinkEvent, ownerGeneration: Long): OfferVerdict {
        if (terminal) return OfferVerdict.TerminalFailure
        if (event.peer.relation.generation != ownerGeneration) return OfferVerdict.TerminalFailure
        if (queue.size >= capacity) return OfferVerdict.Backpressure
        queue.addLast(event)
        // revalidate the owner token upon completion: a mid-offer terminal turn undoeth the slip
        if (terminal) {
            queue.remove(event)
            return OfferVerdict.TerminalFailure
        }
        return OfferVerdict.Accepted
    }

    /** Take the head event for delivery, or null when empty or terminal. */
    fun poll(): LinkEvent? {
        if (terminal) return null
        return queue.pollFirst()
    }

    /** Fail the channel terminally: nothing more is carried, and the queue is voided. */
    fun failTerminally() {
        terminal = true
        queue.clear()
    }
}
